# Лабораторная работа №2. Структуры данных

## Кортежи

### Примеры кортежей

In [3]:
() # Пустой кортеж

()

In [5]:
favoritelang = ("python", "julia", "R") #Кортеж из элементов типа String

("python", "julia", "R")

In [13]:
x1 = (1, 2, 3) # кортеж из целых чисел

(1, 2, 3)

In [14]:
x2 = (1, 2.0, "tmp") # Кортеж из элементов разных типов

(1, 2.0, "tmp")

In [15]:
x3 = (a=2, b=1+2) # Именованный кортеж

(a = 2, b = 3)

### Примеры операций над кортежами

In [16]:
length(x2) # длина кортежа x2

3

In [17]:
x2[1], x2[2], x2[3] # обратиться к элементам кортежа x2

(1, 2.0, "tmp")

In [18]:
с = x1[2] + x1[3] # произвести какую-либо операцию (сложение)
# с вторым и третьим элементами кортежа x1

5

In [19]:
x3.a, x3.b, x3[2] # обращение к элементам именованного кортежа x3

(2, 3, 3)

In [21]:
# проверка вхождения элементов tmp и 0 в кортеж x2
# (два способа обращения к методу in()):
in("tmp", x2), 0 in x2

(true, false)

## Словари

### Примеры словарей и операций над ними

In [23]:
# создать словарь с именем phonebook:
phonebook = Dict("Иванов И.И." => ("867-5309","333-5544"), "Бухгалтерия" => "555-2368")

Dict{String, Any} with 2 entries:
  "Бухгалтерия" => "555-2368"
  "Иванов И.И." => ("867-5309", "333-5544")

In [24]:
# вывести ключи словаря:
keys(phonebook)

KeySet for a Dict{String, Any} with 2 entries. Keys:
  "Бухгалтерия"
  "Иванов И.И."

In [25]:
# вывести значения элементов словаря:
values(phonebook)

ValueIterator for a Dict{String, Any} with 2 entries. Values:
  "555-2368"
  ("867-5309", "333-5544")

In [26]:
# вывести заданные в словаре пары "ключ - значение":
pairs(phonebook)

Dict{String, Any} with 2 entries:
  "Бухгалтерия" => "555-2368"
  "Иванов И.И." => ("867-5309", "333-5544")

In [29]:
# проверка вхождения ключа в словарь:
haskey(phonebook, "Иванов И.И.")

true

In [30]:
# добавить элемент в словарь:
phonebook["Сидоров П.С."] = "555-3344"


"555-3344"

In [31]:
# удалить ключ и связанные с ним значения из словаря
pop!(phonebook, "Иванов И.И.")

("867-5309", "333-5544")

In [32]:
keys(phonebook)

KeySet for a Dict{String, Any} with 2 entries. Keys:
  "Сидоров П.С."
  "Бухгалтерия"

In [33]:
# Объединение словарей (функция merge()):
a = Dict("foo" => 0.0, "bar" => 42.0);
b = Dict("baz" => 17, "bar" => 13.0);
merge(a, b), merge(b,a)


(Dict{String, Real}("bar" => 13.0, "baz" => 17, "foo" => 0.0), Dict{String, Real}("bar" => 42.0, "baz" => 17, "foo" => 0.0))

## Множества

### Примеры множеств и операций над ними

In [34]:
# создать множество из четырёх целочисленных значений:
A = Set([1, 3, 4, 5])


Set{Int64} with 4 elements:
  5
  4
  3
  1

In [35]:
# создать множество из 11 символьных значений:
B = Set("abrakadabra")

Set{Char} with 5 elements:
  'a'
  'd'
  'r'
  'k'
  'b'

In [40]:
# проверка эквивалентности двух множеств:
S1 = Set([1,2]);
S2 = Set([3,4]);
issetequal(S1,S2)

false

In [41]:
# проверка эквивалентности двух множеств:
S3 = Set([1,2,2,3,1,2,3,2,1]);
S4 = Set([2,3,1]);
issetequal(S3,S4)

true

In [42]:
# объединение множеств:
C=union(S1,S2)

Set{Int64} with 4 elements:
  4
  2
  3
  1

In [43]:
# пересечение множеств:
D = intersect(S1,S3)

Set{Int64} with 2 elements:
  2
  1

In [44]:
# разность множеств:
E = setdiff(S3,S1)

Set{Int64} with 1 element:
  3

In [45]:
# проверка вхождения элементов одного множества в другое:
issubset(S1,S4)

true

In [46]:
# добавление элемента в множество:
push!(S4, 99)


Set{Int64} with 4 elements:
  2
  99
  3
  1

In [47]:
# удаление последнего элемента множества:
pop!(S4)

2

In [48]:
S4

Set{Int64} with 3 elements:
  99
  3
  1

## Массивы

### Примеры массивов

In [49]:
# создание пустого массива с абстрактным типом:
empty_array_1 = []

Any[]

In [52]:
# создание пустого массива с конкретным типом:
empty_array_2 = (Int64)[]

Int64[]

In [51]:
empty_array_3 = (Float64)[]

Float64[]

In [53]:
# вектор-столбец:
a = [1, 2, 3]

3-element Vector{Int64}:
 1
 2
 3

In [55]:
# вектор-строка:
b = [1 2 3]

1×3 Matrix{Int64}:
 1  2  3

In [57]:
# многомерные массивы (матрицы):
A = [[1, 2, 3] [4, 5, 6] [7, 8, 9]]

3×3 Matrix{Int64}:
 1  4  7
 2  5  8
 3  6  9

In [58]:
B = [[1 2 3]; [4 5 6]; [7 8 9]]

3×3 Matrix{Int64}:
 1  2  3
 4  5  6
 7  8  9

In [59]:
# одномерный массив из 8 элементов (массив $1 \times 8$)
# со значениями, случайно распределёнными на интервале [0, 1):
c = rand(1,8)

1×8 Matrix{Float64}:
 0.310043  0.950794  0.915795  0.962301  …  0.185676  0.892984  0.766257

In [66]:
# многомерный массив $2 \times 3$ (2 строки, 3 столбца) элементов
# со значениями, случайно распределёнными на интервале [0, 1):
C = rand(2,3)

2×3 Matrix{Float64}:
 0.203864  0.758838  0.393973
 0.738799  0.988954  0.64076

In [67]:
# трёхмерный массив:
D = rand(4, 3, 2)

4×3×2 Array{Float64, 3}:
[:, :, 1] =
 0.227332  0.844477   0.464529
 0.708991  0.0160303  0.854196
 0.760585  0.115136   0.984205
 0.283116  0.597728   0.772792

[:, :, 2] =
 0.0107221  0.446021  0.0808428
 0.472339   0.710554  0.627792
 0.337875   0.181368  0.0408687
 0.903711   0.882048  0.893641

### Примеры массивов, заданных некоторыми функциями через включение

In [68]:
# массив из квадратных корней всех целых чисел от 1 до 10:
roots = [sqrt(i) for i in 1:10]

10-element Vector{Float64}:
 1.0
 1.4142135623730951
 1.7320508075688772
 2.0
 2.23606797749979
 2.449489742783178
 2.6457513110645907
 2.8284271247461903
 3.0
 3.1622776601683795

In [69]:
# массив с элементами вида 3*x^2,
# где x - нечётное число от 1 до 9 (включительно)
ar_1 = [3*i^2 for i in 1:2:9]

5-element Vector{Int64}:
   3
  27
  75
 147
 243

In [70]:
# массив квадратов элементов, если квадрат не делится на 5 или 4:
ar_2=[i^2 for i=1:10 if (i^2%5!=0 && i^2%4!=0)]

4-element Vector{Int64}:
  1
  9
 49
 81

### Некоторые операции для работы с массивами

In [71]:
# одномерный массив из пяти единиц:
ones(5)

5-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0
 1.0

In [73]:
# двумерный массив 2x3 из единиц:
ones(2,3)

2×3 Matrix{Float64}:
 1.0  1.0  1.0
 1.0  1.0  1.0

In [74]:
# одномерный массив из 4 нулей:
zeros(4)

4-element Vector{Float64}:
 0.0
 0.0
 0.0
 0.0

In [95]:
# заполнить массив 3x2 цифрами 3.5
fill(3.5,(3,2))

3×2 Matrix{Float64}:
 3.5  3.5
 3.5  3.5
 3.5  3.5

In [96]:
# заполнение массива посредством функции repeat():
repeat([1,2],3,3)
repeat([1 2],3,3)

3×6 Matrix{Int64}:
 1  2  1  2  1  2
 1  2  1  2  1  2
 1  2  1  2  1  2

In [81]:
# преобразование одномерного массива из целых чисел от 1 до 12
# в двумерный массив 2x6
a = collect(1:12)
b = reshape(a,(2,6))


2×6 Matrix{Int64}:
 1  3  5  7   9  11
 2  4  6  8  10  12

In [83]:
# транспонирование
b'


6×2 adjoint(::Matrix{Int64}) with eltype Int64:
  1   2
  3   4
  5   6
  7   8
  9  10
 11  12

In [84]:
# транспонирование
с = transpose(b)

6×2 transpose(::Matrix{Int64}) with eltype Int64:
  1   2
  3   4
  5   6
  7   8
  9  10
 11  12

In [85]:
# массив 10x5 целых чисел в диапазоне [10, 20]:
ar = rand(10:20, 10, 5)

10×5 Matrix{Int64}:
 15  19  16  16  13
 16  13  17  20  16
 13  16  14  14  11
 15  14  17  12  10
 18  14  12  17  20
 17  18  17  19  16
 13  10  11  19  15
 14  11  19  20  20
 12  11  18  18  13
 13  10  13  10  20

In [86]:
# выбор всех значений строки в столбце 2:
ar[:, 2]


10-element Vector{Int64}:
 19
 13
 16
 14
 14
 18
 10
 11
 11
 10

In [87]:
# выбор всех значений в столбцах 2 и 5:
ar[:, [2, 5]]


10×2 Matrix{Int64}:
 19  13
 13  16
 16  11
 14  10
 14  20
 18  16
 10  15
 11  20
 11  13
 10  20

In [88]:
# все значения строк в столбцах 2, 3 и 4:
ar[:, 2:4]


10×3 Matrix{Int64}:
 19  16  16
 13  17  20
 16  14  14
 14  17  12
 14  12  17
 18  17  19
 10  11  19
 11  19  20
 11  18  18
 10  13  10

In [89]:
# значения в строках 2, 4, 6 и в столбцах 1 и 5:
ar[[2, 4, 6], [1, 5]]

3×2 Matrix{Int64}:
 16  16
 15  10
 17  16

In [90]:
# значения в строке 1 от столбца 3 до последнего столбца:
ar[1, 3:end]

3-element Vector{Int64}:
 16
 16
 13

In [91]:
# сортировка по столбцам:
sort(ar,dims=1)


10×5 Matrix{Int64}:
 12  10  11  10  10
 13  10  12  12  11
 13  11  13  14  13
 13  11  14  16  13
 14  13  16  17  15
 15  14  17  18  16
 15  14  17  19  16
 16  16  17  19  20
 17  18  18  20  20
 18  19  19  20  20

In [92]:
# сортировка по строкам:
sort(ar,dims=2)

10×5 Matrix{Int64}:
 13  15  16  16  19
 13  16  16  17  20
 11  13  14  14  16
 10  12  14  15  17
 12  14  17  18  20
 16  17  17  18  19
 10  11  13  15  19
 11  14  19  20  20
 11  12  13  18  18
 10  10  13  13  20

In [93]:
# поэлементное сравнение с числом
# (результат - массив логических значений):
ar .> 14


10×5 BitMatrix:
 1  1  1  1  0
 1  0  1  1  1
 0  1  0  0  0
 1  0  1  0  0
 1  0  0  1  1
 1  1  1  1  1
 0  0  0  1  1
 0  0  1  1  1
 0  0  1  1  0
 0  0  0  0  1

In [94]:
# возврат индексов элементов массива, удовлетворяющих условию:
findall(ar .> 14)


27-element Vector{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(2, 1)
 CartesianIndex(4, 1)
 CartesianIndex(5, 1)
 CartesianIndex(6, 1)
 CartesianIndex(1, 2)
 CartesianIndex(3, 2)
 CartesianIndex(6, 2)
 CartesianIndex(1, 3)
 CartesianIndex(2, 3)
 CartesianIndex(4, 3)
 CartesianIndex(6, 3)
 CartesianIndex(8, 3)
 ⋮
 CartesianIndex(2, 4)
 CartesianIndex(5, 4)
 CartesianIndex(6, 4)
 CartesianIndex(7, 4)
 CartesianIndex(8, 4)
 CartesianIndex(9, 4)
 CartesianIndex(2, 5)
 CartesianIndex(5, 5)
 CartesianIndex(6, 5)
 CartesianIndex(7, 5)
 CartesianIndex(8, 5)
 CartesianIndex(10, 5)

# Самостоятельная работа

### 1. Даны множества: 𝐴 = {0, 3, 4, 9}, 𝐵 = {1, 3, 4, 7}, 𝐶 = {0, 1, 2, 4, 7, 8, 9}. Найти 𝑃 = 𝐴 ∩ 𝐵 ∪ 𝐴 ∩ 𝐵 ∪ 𝐴 ∩ 𝐶 ∪ 𝐵 ∩ 𝐶

In [102]:
A = Set([0, 3, 4, 9])
B = Set([1, 3, 4, 7])
C = Set([0, 1, 2, 4, 7, 8, 9])
P = union(intersect(A,B), intersect(A,B),intersect(A,C),intersect(B,C))
print(P)

Set([0, 4, 7, 9, 3, 1])

### 2. Приведите свои примеры с выполнением операций над множествами элементов разных типов.


In [104]:
N = Set([2002, 2003, 2004, 2005]) # выпускники кружка Оленьи рога
T = Set([2002, 2000, 2001, 2005, 1998, 1997, 1930, 2009, 1992, 1993, 2003, 2004]) # выпускники школы творчества
# входят ли выпускники кружка Оленьи рога в список выпускников школы творчества
issubset(N,T)

true

In [107]:
N = Set([2002, 2003, 2004, 2005]) # выпускники кружка Оленьи рога
S = Set([2000, 1993, 1994, 1995]) # выпускники кружка Оленьи копыта
R = Set([1930, 2009, 2004, 2005]) # выпускники кружка Оленьи хвостики
T = union(N,S,R) # выпускники школы творчества по годам
print(T)

Set([2004, 2005, 1995, 2002, 1930, 1994, 2003, 1993, 2009, 2000])

### 3. Создайте разными способами:
### 3.1) массив (1, 2, 3, … 𝑁 − 1, 𝑁 ), 𝑁 выберите больше 20;

In [118]:
N = 29
Arr1 = collect(1:1:N)
print(Arr1)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]

### 3.2) массив (𝑁, 𝑁 − 1 … , 2, 1), 𝑁 выберите больше 20;

In [120]:
N = 29
Arr2 = collect(N:-1:1)
print(Arr2)

[29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

### 3.3) массив (1, 2, 3, … , 𝑁 − 1, 𝑁, 𝑁 − 1, … , 2, 1), 𝑁 выберите больше 20;

In [130]:
N = 20
Arr3 = cat(collect(1:1:N),collect(N:-1:1),dims=1)
print(Arr3)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

### 3.4) массив с именем tmp вида (4, 6, 3);

In [133]:
tmp = [4, 6, 3]
print("tmp = ", tmp)

tmp = [4, 6, 3]

### 3.5) массив, в котором первый элемент массива tmp повторяется 10 раз;


In [136]:
Arr3_5 = fill(tmp[1], 10)
print(Arr3_5)

[4, 4, 4, 4, 4, 4, 4, 4, 4, 4]

### 3.6) массив, в котором все элементы массива tmp повторяются 10 раз;


In [137]:
Arr3_6 = repeat(tmp, 10)
print(Arr3_6)

[4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3, 4, 6, 3]

### 3.7) массив, в котором первый элемент массива tmp встречается 11 раз, второй элемент — 10 раз, третий элемент — 10 раз;

In [140]:
Arr3_7 = cat(fill(tmp[1],11),fill(tmp[2],10),fill(tmp[3],10), dims = 1)
print(Arr3_7)

[4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

### 3.8) массив, в котором первый элемент массива tmp встречается 10 раз подряд, второй элемент — 20 раз подряд, третий элемент — 30 раз подряд;

In [141]:
Arr3_8 = cat(fill(tmp[1],10),fill(tmp[2],20),fill(tmp[3],30), dims = 1)
print(Arr3_8)

[4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

### 3.9) массив из элементов вида 2^𝑡𝑚𝑝[𝑖], 𝑖 = 1, 2, 3, где элемент 2^𝑡𝑚𝑝[3] встречается 4 раза; посчитайте в полученном векторе, сколько раз встречается цифра 6, и выведите это значение на экран;

In [146]:
#создали массив
Arr3_9_1 = [2^tmp[i] for i in 1:1:3]
Arr3_9 = cat(Arr3_9_1, repeat([2^tmp[3]],3), dims = 1)
print(Arr3_9)

[16, 64, 8, 8, 8, 8]

In [162]:
# переводим в строку, сплитуем
string_split = join(Arr3_9, "")
print(string_split)

16648888

In [161]:
res = count(x -> x == '6', string_split)
print("Цифра 6 встречается ", res, " раз")

Цифра 6 встречается 2 раз

### 3.10) вектор значений 𝑦 = 𝑒^𝑥 cos(𝑥) в точках 𝑥 = 3, 3.1, 3.2, … , 6, найдите среднее значение 𝑦;

In [184]:
using Statistics
y = [exp(x) * cos(x) for x in 3:0.1:6]
println(mean(y))
   

53.11374594642971


### 3.11) вектор вида (𝑥^𝑖, 𝑦^𝑗), 𝑥 = 0.1, 𝑖 = 3, 6, 9, … , 36, 𝑦 = 0.2, 𝑗 = 1, 4, 7, … , 34;

In [189]:
x = 0.1
y = 0.2
vec3_11 = [(x^i, y^j) for i = 3:3:36, j  = 1:3:34]
println(vec3_11)

[(0.0010000000000000002, 0.2) (0.0010000000000000002, 0.0016000000000000003) (0.0010000000000000002, 1.2800000000000005e-5) (0.0010000000000000002, 1.0240000000000006e-7) (0.0010000000000000002, 8.192000000000005e-10) (0.0010000000000000002, 6.5536000000000055e-12) (0.0010000000000000002, 5.2428800000000056e-14) (0.0010000000000000002, 4.194304000000005e-16) (0.0010000000000000002, 3.3554432000000048e-18) (0.0010000000000000002, 2.684354560000004e-20) (0.0010000000000000002, 2.1474836480000035e-22) (0.0010000000000000002, 1.717986918400003e-24); (1.0000000000000004e-6, 0.2) (1.0000000000000004e-6, 0.0016000000000000003) (1.0000000000000004e-6, 1.2800000000000005e-5) (1.0000000000000004e-6, 1.0240000000000006e-7) (1.0000000000000004e-6, 8.192000000000005e-10) (1.0000000000000004e-6, 6.5536000000000055e-12) (1.0000000000000004e-6, 5.2428800000000056e-14) (1.0000000000000004e-6, 4.194304000000005e-16) (1.0000000000000004e-6, 3.3554432000000048e-18) (1.0000000000000004e-6, 2.68435456000000

### 3.12)

In [191]:
M = 25
vec3_12 = [2^i/i for i = 1:1:M]
print(vec3_12)

[2.0, 2.0, 2.6666666666666665, 4.0, 6.4, 10.666666666666666, 18.285714285714285, 32.0, 56.888888888888886, 102.4, 186.1818181818182, 341.3333333333333, 630.1538461538462, 1170.2857142857142, 2184.5333333333333, 4096.0, 7710.117647058823, 14563.555555555555, 27594.105263157893, 52428.8, 99864.38095238095, 190650.18181818182, 364722.0869565217, 699050.6666666666, 1.34217728e6]

### 3.13)

In [192]:
N = 30
vec3_13 = ["fn$i" for i = 1:1:N]
print(vec3_13)

["fn1", "fn2", "fn3", "fn4", "fn5", "fn6", "fn7", "fn8", "fn9", "fn10", "fn11", "fn12", "fn13", "fn14", "fn15", "fn16", "fn17", "fn18", "fn19", "fn20", "fn21", "fn22", "fn23", "fn24", "fn25", "fn26", "fn27", "fn28", "fn29", "fn30"]

### 3.14) 

In [201]:
x = rand(0:999, 250)
y = rand(0:999, 250)

task1 = [y[i]-x[i-1] for i=2:1:length(y)]
print(task1)

[839, 628, 339, -237, -615, 433, -361, 439, -457, 131, 139, -96, -284, -327, 228, -175, -334, 663, 299, 74, -570, -189, 586, -95, -259, 220, 644, -280, 57, 544, 44, -147, -567, -172, 512, -319, 326, 46, 835, 675, -555, -464, -156, -684, 394, 237, 65, -546, -309, 743, -92, 864, 281, 361, -158, -541, -473, 176, 591, -64, 441, -442, -75, -687, -531, 252, -662, 281, 432, 385, 141, 829, 67, 619, -697, 182, 30, 183, -3, 227, -21, 16, -86, -820, 597, -17, -23, -8, -272, 850, -263, -60, 43, 111, 267, 550, 219, 706, 695, -814, 560, 451, 27, 93, 445, -405, 461, -889, 29, 417, 104, -228, 838, 369, -635, 11, -129, 43, 64, 593, -73, -207, 57, -253, 92, -52, 139, -110, -203, 260, 549, 300, -446, 358, 679, 746, 512, 118, -88, -67, 34, -252, 92, 207, 748, 543, 296, 467, -40, -399, -47, -349, 248, -445, 147, -133, 209, 495, -19, -227, 120, -641, -140, -86, 389, -620, 308, -502, 278, 538, -122, -502, 442, 442, 360, 181, -117, -901, -472, -87, -291, -316, 171, 604, 57, -815, 218, 177, 612, -450, 803, 472

In [202]:
task2 = [x[i]+2*x[i+1]-x[i+3] for i=1:1:length(x)-3]
print(task2)

[-730, -14, 1724, 2279, 1933, 743, 289, 1081, 1113, 1092, 1153, 1404, 1160, 865, 1793, 1602, 571, 342, 396, 1142, 1361, 228, 158, 1646, 1366, 660, 1016, 596, 458, 888, 1211, 1978, 1434, 985, 2112, 1807, 551, -364, -350, 149, 641, 2409, 2068, 466, 360, 728, 2341, 2086, 778, 408, 186, 334, 786, 1606, 1608, 1815, 2036, 1078, -60, -242, 338, 1326, 2247, 1183, 589, 1629, 1298, 771, 1476, 1620, 383, -832, -335, 1750, 1649, 190, 1210, 2493, 1520, 255, -345, 645, 1856, 840, 385, 1180, 2335, 2023, 613, 330, 312, 588, 1266, 1188, 697, 518, -475, 393, 1649, 456, 667, 1983, 1046, 524, 760, 1083, 2070, 870, -171, 1579, 2493, 455, -627, 704, 1840, 2658, 2528, 237, -748, 1304, 2446, 1575, 845, 273, 378, 766, 742, 794, 289, -225, 487, 1191, 946, 226, 275, 171, 531, 1440, 657, 769, 1497, 1873, 1457, 68, 738, 1009, -211, 934, 2161, 1219, 794, 1371, 1195, 1481, 1705, 517, -799, -231, 1154, 1250, 2038, 2537, 957, 461, 1189, 1167, 1560, 1649, 215, 430, 1614, 1293, 810, -203, -517, 999, 2402, 1082, 407, 171

In [204]:
n=250
res = 0
for i in 1:n-1
   res += exp(-x[i+1])/(x[i]+10)
end
print(res)

0.0040356463379898405

In [210]:
for i in 1:n
    if y[i]>600 
        print("y[",i,"]=",y[i],", ")
    end   
end

y[1]=613, y[2]=885, y[3]=684, y[4]=791, y[5]=651, y[7]=937, y[12]=857, y[16]=738, y[19]=692, y[20]=769, y[24]=893, y[27]=757, y[28]=892, y[31]=892, y[32]=752, y[36]=897, y[38]=830, y[40]=916, y[41]=694, y[44]=832, y[47]=869, y[51]=744, y[53]=916, y[54]=894, y[55]=789, y[56]=786, y[59]=878, y[60]=879, y[62]=641, y[64]=645, y[69]=736, y[70]=862, y[71]=920, y[72]=685, y[73]=853, y[75]=828, y[77]=938, y[79]=908, y[80]=929, y[81]=650, y[86]=930, y[88]=760, y[89]=776, y[91]=866, y[95]=614, y[96]=750, y[97]=714, y[99]=820, y[100]=903, y[102]=697, y[103]=978, y[104]=829, y[105]=617, y[108]=757, y[111]=635, y[112]=856, y[113]=692, y[114]=981, y[117]=979, y[118]=768, y[119]=890, y[121]=656, y[122]=745, y[123]=771, y[132]=626, y[133]=639, y[135]=626, y[136]=891, y[137]=868, y[138]=978, y[140]=795, y[144]=634, y[145]=857, y[146]=779, y[147]=928, y[148]=940, y[150]=624, y[154]=924, y[161]=726, y[162]=742, y[164]=807, y[166]=656, y[168]=785, y[170]=863, y[171]=803, y[174]=667, y[175]=823, y[182]=703

In [215]:
for i in 1:n
    if y[i]>600 
        for j in 1:n
            if x[j]==y[i]
                print("y[",i,"]=",y[i],"=x[",j,"]=",x[j], "\n")
            end
        end
    end   
end

y[1]=613=x[53]=613
y[24]=893=x[219]=893
y[32]=752=x[111]=752
y[36]=897=x[117]=897
y[70]=862=x[75]=862
y[71]=920=x[112]=920
y[75]=828=x[49]=828
y[80]=929=x[190]=929
y[81]=650=x[144]=650
y[89]=776=x[250]=776
y[99]=820=x[142]=820
y[103]=978=x[122]=978
y[105]=617=x[215]=617
y[112]=856=x[44]=856
y[114]=981=x[150]=981
y[121]=656=x[16]=656
y[138]=978=x[122]=978
y[147]=928=x[178]=928
y[148]=940=x[48]=940
y[166]=656=x[16]=656
y[170]=863=x[238]=863
y[174]=667=x[156]=667
y[193]=944=x[55]=944
y[197]=863=x[238]=863
y[201]=720=x[63]=720
y[202]=882=x[211]=882
y[203]=764=x[226]=764
y[213]=856=x[44]=856
y[223]=654=x[185]=654
y[226]=827=x[234]=827
y[236]=940=x[48]=940
y[243]=656=x[16]=656
y[246]=862=x[75]=862


### 4. Создайте массив squares, в котором будут храниться квадраты всех целых чисел от 1 до 100.


In [217]:
squares = [i^2 for i in 1:1:100]
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000]

### 5. Подключите пакет Primes (функции для вычисления простых чисел). Сгенерируйте массив myprimes, в котором будут храниться первые 168 простых чисел. Определите 89-е наименьшее простое число. Получите срез массива с 89-го до 99-го элемента включительно, содержащий наименьшие простые числа.

In [225]:
using Primes

# Создаем список простых чисел до 1000 (или другого достаточно большого числа)
myprimes = primes(1000)

println(myprimes[89])
println(myprimes[89:99])

461
[461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523]


### 6.1)


In [221]:
for i in 10:1:100
    res += i^3+4*i^2
end
print(res)

5.3705470004035644e7

### 6.2)


In [226]:
sum2 = sum((2^i / i + 3^i / i^2) for i in 1:25)
println(sum2)

2.1291704368143802e9


### 6.3)


In [227]:
sum3 = sum(prod(2:2n) / prod(3:2n+1) for n in 1:19)
println(sum3)

12.84175745993532
